contract_comm/evm.go

```diff
@@ -0,0 +1,231 @@
+		for header := chain.GetHeader(ref.ParentHash, ref.Number.Uint64()-1); header != nil; header = chain.GetHeader(header.ParentHash, header.Number.Uint64()-1) {
+			cache[header.Number.Uint64()-1] = header.ParentHash
+			if n == header.Number.Uint64()-1 {
+				return header.ParentHash
+			}
+		}
+		return common.Hash{}
+	}
+}
+
+// CanTransfer checks whether there are enough funds in the address' account to make a transfer.
+// This does not take the necessary gas in to account to make the transfer valid.
+func CanTransfer(db vm.StateDB, addr common.Address, amount *big.Int) bool {
+	return db.GetBalance(addr).Cmp(amount) >= 0
+}
+
+// Transfer subtracts amount from sender and adds amount to recipient using the given Db
+func Transfer(db vm.StateDB, sender, recipient common.Address, amount *big.Int) {
+	db.SubBalance(sender, amount)
+	db.AddBalance(recipient, amount)
+}
+
+// An EVM handler to make calls to smart contracts from within geth
+type InternalEVMHandler struct {
+	chain ChainContext
+}
+
+func MakeStaticCall(scRegistryId string, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, header *types.Header, state *state.StateDB) (uint64, error) {
+	scAddress, err := GetContractAddress(scRegistryId, header, state)
+	if err == errors.ErrSmartContractNotDeployed || err == errors.ErrRegistryContractNotDeployed {
+		log.Warn("Contract deployment in progress")
+		return 0, err
+	}
+	if err != nil {
+		return 0, err
+	}
+	return executeEVMFunction(*scAddress, abi, funcName, args, returnObj, gas, nil, header, state, false)
+}
+
+func MakeCall(scRegistryId string, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int, header *types.Header, state *state.StateDB) (uint64, error) {
+	scAddress, err := GetContractAddress(scRegistryId, header, state)
+	if err == errors.ErrSmartContractNotDeployed || err == errors.ErrRegistryContractNotDeployed {
+		log.Warn("Contract deployment in progress")
+		return 0, err
+	}
+	if err != nil {
+		return 0, err
+	}
+	return executeEVMFunction(*scAddress, abi, funcName, args, returnObj, gas, value, header, state, true)
+}
+
+func MakeStaticCallWithAddress(scAddress common.Address, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, header *types.Header, state *state.StateDB) (uint64, error) {
+	return executeEVMFunction(scAddress, abi, funcName, args, returnObj, gas, nil, header, state, false)
+}
+
+func MakeCallWithAddress(scAddress common.Address, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int, header *types.Header, state *state.StateDB) (uint64, error) {
+	return executeEVMFunction(scAddress, abi, funcName, args, returnObj, gas, value, header, state, true)
+}
+
+func GetContractAddress(registryId string, header *types.Header, state *state.StateDB) (*common.Address, error) {
+	vmevm, err := createVMEVM(header, state)
```

Comment:
So this is an ask about the whole PR, but with this change, how do you feel about treating abbreviations in variable names as whole chunks, so going for something like `createVmEvm` instead? Especially since this is the convention used elsewhere.

Reply:
Are you sure about this being the convention? Looks like `NewEVM` is function naming from the core team. I also like `NewEvm` better, but don't want to introduce unnecessary conflicts with upstream geth.